File: packages/java/cnos-java/src/main/java/ai/kitsy/cnos/internal/VaultResolver.java
package ai.kitsy.cnos.internal;

import ai.kitsy.cnos.CnosError;
import ai.kitsy.cnos.VaultAuthConfig;
import ai.kitsy.cnos.VaultDefinition;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Resolves vault authentication credentials and local vault secrets.
 * Mirrors Go's vault_provider.go resolveVaultAuth and secrets.go resolveLocalVaultKey logic.
 */
public final class VaultResolver {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final int KEY_LENGTH = 32;

    private VaultResolver() {}

    /**
     * Resolves the auth config for a vault, using the auth resolution chain.
     */
    public static VaultAuthConfig resolveVaultAuth(String vaultId, VaultDefinition definition,
            Environment env) throws CnosError {
        VaultDefinition.Auth auth = definition.getAuth();
        String method = auth != null && auth.getMethod() != null ? auth.getMethod().trim() : "";
        if (method.isEmpty()) {
            method = defaultVaultMethod(definition.getProvider());
        }

        Map<String, Object> config = auth != null ? auth.getConfig() : Collections.emptyMap();

        switch (method) {
            case "iam":
            case "environment":
                return VaultAuthConfig.ofMethod(method, config);
            case "token": {
                Optional<String> token = resolveFirstSource(auth != null ? auth.getToken() : null, env);
                if (!token.isPresent()) {
                    throw vaultAuthError(vaultId, auth != null ? auth.getToken() : null);
                }
                return VaultAuthConfig.ofToken(token.get(), config);
            }
        }

        // Fallback: try token sources, then passphrase sources
        if (auth != null && auth.getToken() != null && !auth.getToken().getFrom().isEmpty()) {
            Optional<String> token = resolveFirstSource(auth.getToken(), env);
            if (token.isPresent()) {
                return VaultAuthConfig.ofToken(token.get(), config);
            }
        }

        if (auth != null && auth.getPassphrase() != null && !auth.getPassphrase().getFrom().isEmpty()) {
            Optional<String> passphrase = resolveFirstSource(auth.getPassphrase(), env);
            if (passphrase.isPresent()) {
                return VaultAuthConfig.ofPassphrase(passphrase.get(), config);
            }
            throw vaultAuthError(vaultId, auth.getPassphrase());
        }

        Optional<String> passphrase = resolveVaultPassphrase(vaultId, env);
        if (passphrase.isPresent()) {
            return VaultAuthConfig.ofPassphrase(passphrase.get(), config);
        }

        return VaultAuthConfig.ofMethod(method, config);
    }

    /**
     * Resolves the AES key for a local vault, using the auth chain.
     */
    public static byte[] resolveLocalVaultKey(String secretHome, String vaultId,
            LocalVault.Metadata meta, VaultDefinition definition, Environment env) throws CnosError {

        // 1. Check __CNOS_VAULT_KEY_{VAULT_UPPER}__ env var
        Optional<String> preKey = env.get(getVaultSessionKeyEnvVar(vaultId));
        if (preKey.isPresent() && !preKey.get().isEmpty()) {
            try {
                byte[] key = LocalVault.hexDecode(preKey.get());
                if (key.length == KEY_LENGTH) return key;
            } catch (Exception ignored) {}
        }

        // 2. Check session file
        File sessionFile = new File(secretHome, "sessions/" + vaultId + ".json");
        if (sessionFile.exists()) {
            Optional<byte[]> sessionKey = readSessionKey(sessionFile);
            if (sessionKey.isPresent()) return sessionKey.get();
        }

        // 3. Walk auth sources
        for (String source : resolveLocalVaultAuthSources(vaultId, definition)) {
            if (source.startsWith("env:")) {
                String varName = source.substring(4);
                Optional<String> passphrase = env.get(varName);
                if (passphrase.isPresent() && !passphrase.get().isEmpty()) {
                    return LocalVault.deriveKey(passphrase.get(), meta.salt, meta.iterations);
                }
            } else if (source.startsWith("keychain:")) {
                // OS keychain is not reliably available cross-platform in Java — skip
            } else if ("prompt".equals(source)) {
                // Interactive prompt: try console if available
                java.io.Console console = System.console();
                if (console != null) {
                    char[] pw = console.readPassword("Enter passphrase for vault \"%s\": ", vaultId);
                    if (pw != null && pw.length > 0) {
                        String passphrase = new String(pw);
                        java.util.Arrays.fill(pw, '\0');
                        return LocalVault.deriveKey(passphrase, meta.salt, meta.iterations);
                    }
                }
            }
        }

        // 4. Fallback env var resolution
        Optional<String> passphrase = resolveVaultPassphrase(vaultId, env);
        if (passphrase.isPresent()) {
            return LocalVault.deriveKey(passphrase.get(), meta.salt, meta.iterations);
        }

        String keyVar = getVaultSessionKeyEnvVar(vaultId);
        List<String> tried = new ArrayList<>();
        tried.add(keyVar);
        tried.addAll(resolveLocalVaultAuthSources(vaultId, definition));
        throw new CnosError("cnos: cannot authenticate to vault \"" + vaultId
                + "\". Tried: " + String.join(", ", tried)
                + ". Set " + getVaultPassphraseEnvVar(vaultId)
                + " or run cnos vault auth " + vaultId);
    }

    static List<String> resolveLocalVaultAuthSources(String vaultId, VaultDefinition definition) {
        if (definition != null && definition.getAuth() != null
                && definition.getAuth().getPassphrase() != null
                && !definition.getAuth().getPassphrase().getFrom().isEmpty()) {
            return new ArrayList<>(definition.getAuth().getPassphrase().getFrom());
        }

        String token = normalizeVaultToken(vaultId);
        List<String> sources = new ArrayList<>();
        if (!token.isEmpty()) {
            sources.add("env:CNOS_SECRET_PASSPHRASE_" + token);
        }
        sources.add("env:CNOS_SECRET_PASSPHRASE");
        sources.add("keychain:cnos/" + vaultId);
        sources.add("prompt");
        return sources;
    }

    static Optional<String> resolveVaultPassphrase(String vaultId, Environment env) {
        Optional<String> specific = env.get(getVaultPassphraseEnvVar(vaultId));
        if (specific.isPresent() && !specific.get().isEmpty()) return specific;
        Optional<String> fallback = env.get("CNOS_SECRET_PASSPHRASE");
        if (fallback.isPresent() && !fallback.get().isEmpty()) return fallback;
        return Optional.empty();
    }

    static String getVaultPassphraseEnvVar(String vaultId) {
        String token = normalizeVaultToken(vaultId);
        if (!token.isEmpty() && !"DEFAULT".equals(token)) {
            return "CNOS_SECRET_PASSPHRASE_" + token;
        }
        return "CNOS_SECRET_PASSPHRASE";
    }

    static String getVaultSessionKeyEnvVar(String vaultId) {
        String token = normalizeVaultToken(vaultId);
        if (token.isEmpty()) token = "DEFAULT";
        return "__CNOS_VAULT_KEY_" + token + "__";
    }

    /**
     * Normalizes a vault ID for use in environment variable names.
     * Upper-cases letters, replaces non-alphanumeric chars with {@code _}, collapses repeats, trims.
     */
    public static String normalizeVaultToken(String vaultId) {
        if (vaultId == null) return "";
        vaultId = vaultId.trim();
        StringBuilder sb = new StringBuilder();
        boolean lastUnderscore = false;
        for (char c : vaultId.toCharArray()) {
            if (c >= 'a' && c <= 'z') {
                sb.append((char) (c - 32));
                lastUnderscore = false;
            } else if ((c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
                sb.append(c);
                lastUnderscore = false;
            } else {
                if (!lastUnderscore) {
                    sb.append('_');
                    lastUnderscore = true;
                }
            }
        }
        // Trim leading and trailing underscores
        String result = sb.toString();
        int start = 0, end = result.length();
        while (start < end && result.charAt(start) == '_') start++;
        while (end > start && result.charAt(end - 1) == '_') end--;
        return result.substring(start, end);
    }

    public static String defaultVaultMethod(String provider) {
        if ("local".equals(provider)) return "passphrase";
        if ("github-secrets".equals(provider) || "environment".equals(provider)) return "environment";
        return "";
    }

    private static Optional<byte[]> readSessionKey(File sessionFile) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> doc = MAPPER.readValue(sessionFile, Map.class);
            Object versionObj = doc.get("version");
            Object keyObj = doc.get("derivedKey");
            if (!(versionObj instanceof Integer) || (Integer) versionObj != 1) return Optional.empty();
            if (!(keyObj instanceof String) || ((String) keyObj).isEmpty()) return Optional.empty();
            byte[] key = LocalVault.hexDecode((String) keyObj);
            if (key.length == KEY_LENGTH) return Optional.of(key);
        } catch (Exception ignored) {}
        return Optional.empty();
    }

    private static Optional<String> resolveFirstSource(VaultDefinition.AuthSource source, Environment env) {
        if (source == null) return Optional.empty();
        for (String candidate : source.getFrom()) {
            Optional<String> value = resolveVaultSource(candidate.trim(), env);
            if (value.isPresent()) return value;
        }
        return Optional.empty();
    }

    private static Optional<String> resolveVaultSource(String source, Environment env) {
        if (source.startsWith("env:")) {
            Optional<String> v = env.get(source.substring(4));
            if (v.isPresent() && !v.get().trim().isEmpty()) return Optional.of(v.get().trim());
        } else if (source.startsWith("file:")) {
            try {
                File f = new File(expandHome(source.substring(5)));
                String v = new String(Files.readAllBytes(f.toPath())).trim();
                if (!v.isEmpty()) return Optional.of(v);
            } catch (IOException ignored) {}
        }
        // keychain: skip on Java
        return Optional.empty();
    }

    private static CnosError vaultAuthError(String vaultId, VaultDefinition.AuthSource source) {
        List<String> sources = source != null ? source.getFrom() : Collections.emptyList();
        return new CnosError("cnos: cannot authenticate to vault \"" + vaultId
                + "\". Tried: " + String.join(", ", sources));
    }

    static String expandHome(String path) {
        if (path.startsWith("~/") || path.equals("~")) {
            String home = System.getProperty("user.home", "");
            if (path.equals("~")) return home;
            return home + "/" + path.substring(2);
        }
        return path;
    }
}

File: packages/csharp/Kitsy.Cnos/Internal/VaultResolver.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using Kitsy.Cnos;

namespace Kitsy.Cnos.Internal
{
    internal static class VaultResolver
    {
        private const int KeyLength = 32;

        public static VaultAuthConfig ResolveVaultAuth(string vaultId, VaultDefinition definition,
            CnosEnvironment env)
        {
            VaultAuth? auth = definition.Auth;
            string method = auth?.Method?.Trim() ?? "";
            if (string.IsNullOrEmpty(method))
                method = DefaultVaultMethod(definition.Provider);

            var config = auth?.Config;

            switch (method)
            {
                case "iam":
                case "environment":
                    return VaultAuthConfig.OfMethod(method, config);
                case "token":
                {
                    string? token = ResolveFirstSource(auth?.Token, env);
                    if (token == null) throw VaultAuthError(vaultId, auth?.Token);
                    return VaultAuthConfig.OfToken(token, config);
                }
            }

            // Fallback: try token sources, then passphrase sources
            if (auth?.Token?.From?.Count > 0)
            {
                string? token = ResolveFirstSource(auth.Token, env);
                if (token != null) return VaultAuthConfig.OfToken(token, config);
            }

            if (auth?.Passphrase?.From?.Count > 0)
            {
                string? passphrase = ResolveFirstSource(auth.Passphrase, env);
                if (passphrase != null) return VaultAuthConfig.OfPassphrase(passphrase, config);
                throw VaultAuthError(vaultId, auth.Passphrase);
            }

            string? pp = ResolveVaultPassphrase(vaultId, env);
            if (pp != null) return VaultAuthConfig.OfPassphrase(pp, config);

            return VaultAuthConfig.OfMethod(method, config);
        }

        public static byte[] ResolveLocalVaultKey(string secretHome, string vaultId,
            LocalVault.Metadata meta, VaultDefinition? definition, CnosEnvironment env)
        {
            // 1. Session key env var
            string sessionKeyVar = GetVaultSessionKeyEnvVar(vaultId);
            string? preKey = env.Get(sessionKeyVar);
            if (!string.IsNullOrEmpty(preKey))
            {
                try
                {
                    byte[] k = LocalVault.HexDecode(preKey);
                    if (k.Length == KeyLength) return k;
                }
                catch { }
            }

            // 2. Session file
            string sessionFile = Path.Combine(secretHome, "sessions", vaultId + ".json");
            if (File.Exists(sessionFile))
            {
                byte[]? sk = ReadSessionKey(sessionFile);
                if (sk != null) return sk;
            }

            // 3. Auth sources
            foreach (string source in ResolveLocalVaultAuthSources(vaultId, definition))
            {
                if (source.StartsWith("env:", StringComparison.Ordinal))
                {
                    string? pp = env.Get(source.Substring(4));
                    if (!string.IsNullOrEmpty(pp))
                        return LocalVault.DeriveKey(pp!, meta.Salt, meta.Iterations);
                }
                else if (source == "prompt")
                {
                    // Interactive prompt when running in a terminal
                    if (System.Console.IsInputRedirected) continue;
                    System.Console.Write($"Enter passphrase for vault \"{vaultId}\": ");
                    string? pp = ReadPassword();
                    if (!string.IsNullOrEmpty(pp))
                        return LocalVault.DeriveKey(pp!, meta.Salt, meta.Iterations);
                }
            }

            // 4. Fallback env var
            string? fallback = ResolveVaultPassphrase(vaultId, env);
            if (fallback != null)
                return LocalVault.DeriveKey(fallback, meta.Salt, meta.Iterations);

            var tried = new List<string> { sessionKeyVar };
            tried.AddRange(ResolveLocalVaultAuthSources(vaultId, definition));
            throw new CnosError(
                $"cnos: cannot authenticate to vault \"{vaultId}\". " +
                $"Tried: {string.Join(", ", tried)}. " +
                $"Set {GetVaultPassphraseEnvVar(vaultId)} or run cnos vault auth {vaultId}");
        }

        public static List<string> ResolveLocalVaultAuthSources(string vaultId, VaultDefinition? definition)
        {
            if (definition?.Auth?.Passphrase?.From?.Count > 0)
                return new List<string>(definition.Auth.Passphrase.From);

            string token = NormalizeVaultToken(vaultId);
            var sources = new List<string>();
            if (!string.IsNullOrEmpty(token))
                sources.Add("env:CNOS_SECRET_PASSPHRASE_" + token);
            sources.Add("env:CNOS_SECRET_PASSPHRASE");
            sources.Add("keychain:cnos/" + vaultId);
            sources.Add("prompt");
            return sources;
        }

        public static string? ResolveVaultPassphrase(string vaultId, CnosEnvironment env)
        {
            string? specific = env.Get(GetVaultPassphraseEnvVar(vaultId));
            if (!string.IsNullOrEmpty(specific)) return specific;
            string? fallback = env.Get("CNOS_SECRET_PASSPHRASE");
            if (!string.IsNullOrEmpty(fallback)) return fallback;
            return null;
        }

        public static string GetVaultPassphraseEnvVar(string vaultId)
        {
            string token = NormalizeVaultToken(vaultId);
            if (!string.IsNullOrEmpty(token) && token != "DEFAULT")
                return "CNOS_SECRET_PASSPHRASE_" + token;
            return "CNOS_SECRET_PASSPHRASE";
        }

        public static string GetVaultSessionKeyEnvVar(string vaultId)
        {
            string token = NormalizeVaultToken(vaultId);
            if (string.IsNullOrEmpty(token)) token = "DEFAULT";
            return $"__CNOS_VAULT_KEY_{token}__";
        }

        public static string NormalizeVaultToken(string? vaultId)
        {
            if (string.IsNullOrEmpty(vaultId)) return "";
            var sb = new System.Text.StringBuilder();
            bool lastUnderscore = false;
            foreach (char c in vaultId.Trim())
            {
                if (c >= 'a' && c <= 'z') { sb.Append((char)(c - 32)); lastUnderscore = false; }
                else if ((c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) { sb.Append(c); lastUnderscore = false; }
                else if (!lastUnderscore) { sb.Append('_'); lastUnderscore = true; }
            }
            string result = sb.ToString().Trim('_');
            return result;
        }

        public static string DefaultVaultMethod(string? provider) => provider switch
        {
            "local" => "passphrase",
            "environment" or "github-secrets" => "environment",
            _ => "",
        };

        private static string? ResolveFirstSource(VaultAuthSource? source, CnosEnvironment env)
        {
            if (source?.From == null) return null;
            foreach (string candidate in source.From)
            {
                string? v = ResolveVaultSource(candidate.Trim(), env);
                if (v != null) return v;
            }
            return null;
        }

        private static string? ResolveVaultSource(string source, CnosEnvironment env)
        {
            if (source.StartsWith("env:", StringComparison.Ordinal))
            {
                string? v = env.Get(source.Substring(4));
                if (!string.IsNullOrWhiteSpace(v)) return v!.Trim();
            }
            else if (source.StartsWith("file:", StringComparison.Ordinal))
            {
                try
                {
                    string path = ExpandHome(source.Substring(5));
                    string v = File.ReadAllText(path).Trim();
                    if (!string.IsNullOrEmpty(v)) return v;
                }
                catch { }
            }
            return null;
        }

        private static CnosError VaultAuthError(string vaultId, VaultAuthSource? source)
        {
            string tried = source?.From != null ? string.Join(", ", source.From) : "";
            return new CnosError($"cnos: cannot authenticate to vault \"{vaultId}\". Tried: {tried}");
        }

        private static byte[]? ReadSessionKey(string sessionFile)
        {
            try
            {
                using var doc = JsonDocument.Parse(File.ReadAllBytes(sessionFile));
                var root = doc.RootElement;
                if (!root.TryGetProperty("version", out var ver) || ver.GetInt32() != 1) return null;
                if (!root.TryGetProperty("derivedKey", out var dk)) return null;
                string? hex = dk.GetString();
                if (string.IsNullOrEmpty(hex)) return null;
                byte[] k = LocalVault.HexDecode(hex);
                return k.Length == KeyLength ? k : null;
            }
            catch { return null; }
        }

        private static string? ReadPassword()
        {
            var sb = new System.Text.StringBuilder();
            ConsoleKeyInfo key;
            while (true)
            {
                key = System.Console.ReadKey(intercept: true);
                if (key.Key == ConsoleKey.Enter) break;
                if (key.Key == ConsoleKey.Backspace && sb.Length > 0) sb.Length--;
                else sb.Append(key.KeyChar);
            }
            System.Console.WriteLine();
            return sb.Length > 0 ? sb.ToString() : null;
        }

        public static string ExpandHome(string path)
        {
            if (path == "~") return System.Environment.GetFolderPath(System.Environment.SpecialFolder.UserProfile);
            if (path.StartsWith("~/", StringComparison.Ordinal))
                return System.Environment.GetFolderPath(System.Environment.SpecialFolder.UserProfile) + path.Substring(1);
            return path;
        }
    }
}
